File: src/java/leetcode/editor/en/P99_RecoverBinarySearchTree.java
package java.leetcode.editor.en;
import java.util.*;
import java.util.function.*;
import java.leetcode.editor.en.definition.*;

public class P99_RecoverBinarySearchTree {
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private TreeNode prev;
    public void recoverTree(TreeNode root) {
        TreeNode prev = null;
        Deque<TreeNode> inverse = new ArrayDeque<>();
        dfs(root, inverse);
        inverse.getFirst().val = swap(inverse.getLast().val, inverse.getLast().val = inverse.getFirst().val);
    }
    private void dfs(TreeNode root, Deque<TreeNode> inverse) {
        if (root == null || inverse.size() == 4) return;
        dfs(root.left, inverse);
        if (prev != null && root.val < prev.val) {
            inverse.add(prev);
            inverse.add(root);
        }
        prev = root;
        dfs(root.right, inverse);
    }
    private <T> T swap(T second, T dontcare) {
        return second;
    }
}
//leetcode submit region end(Prohibit modification and deletion)
}

File: src/kt/leetcode/editor/en/P99_RecoverBinarySearchTree.kt
package kt.leetcode.editor.en;
import java.util.*;
import java.util.function.*;
import kt.leetcode.editor.en.definition.*;

public class P99_RecoverBinarySearchTree {
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun recoverTree(root: TreeNode?): Unit {
        var prev: TreeNode? = null
        val inverse = ArrayList<TreeNode>()
        /*  nested function begin */
        fun dfs(root: TreeNode?) {
            if (root == null || inverse.size == 4) return
            dfs(root.left)
            prev?.let {
                if (root.`val` <= it.`val`) {
                    inverse.add(it)
                    inverse.add(root)
                }
            }
            prev = root
            dfs(root.right)
        }
        /*  nested function end */
        dfs(root)
        inverse.first().`val` = inverse.last().`val`.apply { inverse.last().`val` = inverse.first().`val` }
    }
}
//leetcode submit region end(Prohibit modification and deletion)
}
